### PR TITLE
Update integration

### PR DESCRIPTION
**Pull request by augustfe**
## Summary

This pull request implements a major cleanup and refactor of the temporal integration and matrix operator layers, significantly simplifying the codebase and improving maintainability. The changes migrate operator semantics into the matrix classes, unify operator handling, introduce new matrix abstractions, and streamline temporal integrator implementations. Extensive design notes and verification steps are included to document the rationale and ensure correctness. A number of smaller changes improve test performance, update dependencies, and fix minor issues.

**Matrix and Operator Abstraction Improvements:**

* Migrated operator semantics (e.g., diagonal detection, zero/identity handling, structured linear combinations) into matrix classes such as `DiaMatrix`, `Matrix`, and new types like `DiagonalMatrix`, `IdentityMatrix`, and `ZeroMatrix`. This eliminates the need for the `LinearTerm` wrapper and centralizes operator logic.
* Added matrix dunder algebra for cross-type promotion (e.g., `DiaMatrix + Matrix` returns `Matrix`), and removed generic combination factories like `OperatorSum` and `linear_combination()`.
* Introduced `DiagonalMatrix` as a subclass of `DiaMatrix` for pure main-diagonal operators, with elementwise `matvec`/`solve` and arithmetic that preserves diagonal structure.
* Updated ETDRK4 integrator to represent timestep coefficients as matrix objects (`DiagonalMatrix` or `Matrix`), enabling a uniform interface for all operator applications.
* Removed runtime operator dispatch helpers in favor of direct matrix interface calls (`op @ u`, `op.solve(rhs)`, etc.), making temporal integrators simpler and more robust.

**Temporal Integrator Refactoring:**

* Refactored `BaseIntegrator`, `BackwardEuler`, and `ETDRK4` to store and operate directly on normalized matrix-like operators, delegating all operations to matrix methods.
* Simplified ETDRK4 setup and step logic by inlining diagonal coefficient handling and removing redundant branches.
* Improved cache warming for matrix LU factorizations and diagonal solves, ensuring efficient operation inside JAX-jitted paths.

**Testing and Performance:**

* Verified all changes with focused integrator, matrix regression, and operator tests, and documented memory usage and JAX/XLA compilation behavior during test runs.
* Changed default pytest parallelism from "logical" to 2 processes to reduce per-worker memory pressure.
* Updated `ty` dependency to version 0.0.37.

**Examples and API Cleanups:**

* Updated the `examples/helmholtz1D.py` example to use the new Chebyshev basis and test space API, and improved domain handling.
* Added missing or corrected internal fields for `BaseVector` and `BaseDyadic` in `coordinates.py` to match expected usage. [[1]](diffhunk://#diff-9cc04ee435ae81aeda77c7ea49e695c20adeb1e355be7d7b5b95be7620d6c98eR300-R303) [[2]](diffhunk://#diff-9cc04ee435ae81aeda77c7ea49e695c20adeb1e355be7d7b5b95be7620d6c98eR385-R389)
* Added missing import for `OrthogonalSpace` in `ChebyshevU.py`.

For further details, see the extensive design notes and verification steps in the updated `TODO.md`.

## Types of Changes

- [ ] Bug fix
- [x] New feature
- [ ] Performance improvement
- [ ] Documentation update
- [x] Refactor / cleanup
- [ ] Build / CI
- [ ] Other (describe):

## Description of Changes

Cleaned up the integrators to use the new Matrix interface.

## Breaking Changes

- [ ] Yes (describe below)
- [x] No

## Checklist

- [x] Tests added or updated
- [x] Documentation updated (README, examples, docstrings)
- [x] Added type hints
- [x] Linting & formatting pass locally (`pre-commit run --all-files`)
- [x] All new and existing tests pass (`uv run pytest`)
- [ ] Coverage does not decrease
- [ ] Git history is clean (squash/fixup before merge)

### CODE REVIEW

**Comment by augustfe** [comment]
@copilot resolve the merge conflicts in this pull request

File: src/jaxfun/galerkin/arguments.py [hunk 1]
```diff
@@ -762,10 +763,14 @@ def __call__(self, x: Array) -> Array:
...
         """
-        if isinstance(self.functionspace, OrthogonalSpace | DirectSum):
-            return self.functionspace.evaluate(x, self.array)
-        z = self.functionspace.evaluate(x, self.array, True)
-        if self.functionspace.rank == 0:
+        functionspace = self.functionspace
+        if isinstance(functionspace, OrthogonalSpace | DirectSum):
+            return functionspace.evaluate(x, self.array)
+        assert isinstance(
+            functionspace, TensorProductSpace | VectorTensorProductSpace | DirectSumTPS
```
Comment (mikaem) [inline src/jaxfun/galerkin/arguments.py:770]: DirectSumTPS is a TPS. Is this required? We could perhaps have a type for all the multidimensional spaces? And one for 1D spaces. To simplify all the checks here:-)

File: src/jaxfun/galerkin/Chebyshev.py [hunk 2]
```diff
@@ -146,7 +146,7 @@ def quad_points_and_weights(self, N: int | None = None) -> tuple[Array, Array]:
         )
 
     @jit_vmap(in_axes=(0, None), static_argnums=(0, 2))
-    def eval_basis_function(self, X: Array, i: int) -> Array:
+    def eval_basis_function(self, X: float | Array, i: int) -> Array:
```
Comment (mikaem) [inline src/jaxfun/galerkin/Chebyshev.py:149]: This used to be float | Array, right? What changed ty's mind? Since it is vmapped I think it makes most sense to have this only as Array.

File: src/jaxfun/la/diamatrix.py [hunk 3]
```diff
@@ -572,24 +617,36 @@ def lu_solve(
             )
         method = DiaMatrixSolveMethod(method)
 
+        diagonal = self.diagonal_or_none()
+        if diagonal is not None:
+            return _solve_diagonal(diagonal, b, axis)
```
Comment (mikaem) [inline src/jaxfun/la/diamatrix.py:618]: There is actually already a fast path in the regular LUFactors.solve for diagonal matrices. Just look at the top of _forward_elimination and _back_substitution. But I like DiagonalMatrix so never mind. This is probably even faster.

File: src/jaxfun/la/diamatrix.py [hunk 4]
```diff
@@ -1419,11 +1524,158 @@ def rcm(self) -> tuple[DiaMatrix, Array, Array]:
...
+    def diagonal(self, k: int = 0) -> Array:
+        if k == 0:
+            return self.data[0]
+        n = max(self.shape[0] - abs(k), 0)
+        return jnp.zeros(n, dtype=self.data.dtype)
+
+    def diagonal_or_none(self) -> Array:
+        return self.data[0]
+
+    def matvec(self, x: Array, axis: int = 0) -> Array:
```
Comment (mikaem) [inline src/jaxfun/la/diamatrix.py:1511]: rmatvec? It is important. And there is very little jitting here?

File: src/jaxfun/la/operators.py [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+def _dtype_or_default(dtype) -> jnp.dtype:
+    return jnp.dtype(jnp.float32 if dtype is None else dtype)
+
+
+def _check_same_shape(left, right) -> None:
+    if left.shape != right.shape:
+        raise ValueError(f"Shape mismatch: {left.shape} vs {right.shape}")
+
+
+class IdentityMatrix(nnx.Pytree):  # noqa: B903
```
Comment (mikaem) [inline src/jaxfun/la/operators.py]: Would it make more sense to subclass this from DiaMatrix?

File: src/jaxfun/la/operators.py [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+        return NotImplemented
+
+    def __len__(self) -> int:
+        return self.shape[0]
+
+    def __getitem__(self, key: tuple[int, int], /) -> Array:
+        return self.todense()[key]
+
+
+class ZeroMatrix(nnx.Pytree):  # noqa: B903
```
Comment (mikaem) [inline src/jaxfun/la/operators.py]: Same here. I just think it makes sense to have a common baseclass, other than the Pytree.

**Review by mikaem — APPROVED** [review]
Looks great. Big improvement.

File: src/jaxfun/galerkin/Chebyshev.py (hunk 2 above)
Comment (augustfe) [reply src/jaxfun/galerkin/Chebyshev.py:149]: I think it changed something about how it assumes decorators work. It suddenly started complaining about a lot of places (mostly tests if I recall) where the functions were called with floats

File: src/jaxfun/la/diamatrix.py (hunk 4 above)
Comment (augustfe) [reply src/jaxfun/la/diamatrix.py:1511]: Things aren't jitted as I was working my way down from the integrators, where the only jitting is all the way at the top

**Review by mikaem — APPROVED** [review]
😃😃
